File: python/trees.py
"""Traversing trees."""

from __future__ import annotations
from typing import (
    Optional, TypeVar, Generic,
    Callable
)
from dataclasses import dataclass
from enum import Enum

T = TypeVar('T')


@dataclass
class Node(Generic[T]):
    """Inner node of a tree."""

    value: T
    left: Tree[T] = None
    right: Tree[T] = None


# A tree is either a Node or None
Tree = Optional[Node[T]]


tree = Node(
    'A',
    Node('B', Node('D', right=Node('H')), Node('E')),
    Node('C', Node('F'), Node('G', Node('I'), Node('J')))
)


def inorder(t: Tree[T], acc: list[T] | None = None) -> list[T]:
    """Inorder traversal of a tree."""
    acc = acc if acc is not None else []
    if t:
        inorder(t.left, acc)
        acc.append(t.value)
        inorder(t.right, acc)
    return acc


def cps_rec(t: Tree[T], acc: list[T], k: Callable[[], None]) -> None:
    """CPS inorder traversal of a tree."""
    if t is None:
        return k()

    def handle_value() -> None:
        assert t is not None  # For the type-checker
        acc.append(t.value)
        return cps_rec(t.right, acc, k)

    return cps_rec(t.left, acc, handle_value)


def cps(t: Tree[T]) -> list[T]:
    """Inorder traversal of a tree."""
    res: list[T] = []
    cps_rec(t, res, lambda: None)
    return res


class Op(Enum):
    """
    Operations we need to do with objects on the stack.

    These roughly correspond to the continuation points in
    the recursive solution.
    """

    TRAVERSE = 1
    EMIT = 2


def stack_traversal(t: Tree[T]) -> list[T]:
    """Inorder traversal of a tree."""
    stack = [(Op.TRAVERSE, t)]
    res = []
    while stack:
        op, t = stack.pop()
        if op == Op.EMIT:
            assert t is not None
            res.append(t.value)
        else:
            if t:
                stack.append((Op.TRAVERSE, t.right))
                stack.append((Op.EMIT, t))
                stack.append((Op.TRAVERSE, t.left))
    return res

File: go/trees.go
package main

import "fmt"

type node[T any] struct {
	value T
	left  *node[T]
	right *node[T]
}

func cps[T any](t *node[T], acc []T, k func([]T) []T) []T {
	if t == nil {
		return k(acc)
	}

	handleValueAndRight := func(acc []T) []T {
		handleRight := func(acc []T) []T {
			acc = append(acc, t.value)
			return k(acc)
		}
		return cps(t.right, acc, handleRight)
	}

	return cps(t.left, acc, handleValueAndRight)
}

func inorder[T any](t *node[T]) []T {
	return cps(t, []T{}, func(x []T) []T { return x })
}

func main() {
	tree := &node[string]{"A",
		&node[string]{"B",
			&node[string]{"D", nil, &node[string]{"H", nil, nil}},
			&node[string]{"E", nil, nil}},
		&node[string]{"C",
			&node[string]{"F", nil, nil},
			&node[string]{"G", &node[string]{"I", nil, nil}, &node[string]{"J", nil, nil}}},
	}
	fmt.Println(inorder(tree))
}
